functions/savePendingContacts.ts

```diff
@@ -0,0 +1,127 @@
+/* eslint-disable import/no-dynamic-require */
+/* eslint-disable global-require */
+import '@twilio-labs/serverless-runtime-types';
+import {
+  Context,
+  ServerlessCallback,
+  ServerlessFunctionSignature,
+} from '@twilio-labs/serverless-runtime-types/types';
+import { responseWithCors, bindResolve, error500, success } from '@tech-matters/serverless-helpers';
+
+const TokenValidator = require('twilio-flex-token-validator').functionValidator;
+
+type EnvVars = {
+  SYNC_SERVICE_SID: string;
+  SAVE_CONTACT_FN: string;
+};
+
+type Nullable<T> = T | undefined | null;
+type SaveContactFn = (payload: any) => Promise<any>;
+
+/**
+ * Gets a promisefied saveContactFn() given a function name.
+ *
+ * @param functionName Name of the Serverless Function (e.g. 'saveContactToSaferNet')
+ * @param context Context
+ * @param event Event
+ * @returns @type {Nullable<SaveContactFn>} saveContactFn
+ */
+const getSaveContactFn = (
+  functionName: string,
+  context: Context,
+  event: any,
+): Nullable<SaveContactFn> => {
+  const functionPath = Runtime.getFunctions()[functionName]?.path;
+  if (!functionPath) return;
+
+  const saveContactHandler = require(functionPath)?.handler;
+  if (!saveContactHandler) return;
+
+  // eslint-disable-next-line consistent-return
+  return payload =>
+    new Promise((resolveCallback, rejectCallback) => {
+      // Callback passed to saveContactHandler
+      const callback: ServerlessCallback = (error: any, callbackPayload: any) => {
+        const isError = error || ![200, 204].includes(callbackPayload?.statusCode);
+        return isError ? rejectCallback(callbackPayload) : resolveCallback(callbackPayload);
+      };
+
+      return saveContactHandler(context, { ...event, payload }, callback);
+    });
+};
+
+export const handler: ServerlessFunctionSignature = TokenValidator(
+  async (context: Context<EnvVars>, event: any, callback: ServerlessCallback) => {
+    const response = responseWithCors();
+    const resolve = bindResolve(callback)(response);
+
+    const { SYNC_SERVICE_SID, SAVE_CONTACT_FN } = context;
+
+    try {
+      if (!SAVE_CONTACT_FN) throw new Error('SAVE_CONTACT_FN env var not provided.');
+
+      const saveContactFn = getSaveContactFn(SAVE_CONTACT_FN, context, event);
+      if (!saveContactFn) {
+        return resolve(error500(new Error('Could not find a saveContact function'))); // Should it be HTTP 404?
+      }
+
+      const sharedStateClient = context.getTwilioClient().sync.services(SYNC_SERVICE_SID);
+      const list = await sharedStateClient.syncLists('pending-contacts').fetch();
+      const pendingContacts = await list.syncListItems().list();
+
+      type SyncListItemInstance = typeof pendingContacts[0];
+
+      const successfulRetriesListItems: SyncListItemInstance[] = [];
+      const failedRetriesListItems: SyncListItemInstance[] = [];
+
+      await Promise.all(
```

Comment:
Might it be better to perform the whole save -> update / remove pending item operation in one parallelised operation (i.e. one `Promise.all(...)`)?

That might be more efficient that doing all the Save ops in parallel, waiting for them all to finish, then doing all the pending state changes. Also, if one of save ops faIls it won't interfere with the successful saves going on to update their state? Unless I'm missing something about the logic.

Reply:
Yes, the whole flux (saving => updating/deleting) can be done in parallel between each pending contact. When deleting, we use `ListItem.index`, but this `index` is not actually an index, but an id. So it can be done in parallel. I'll make the change.

Reply:
Done